lib/bolt/logger.rb

```diff
@@ -123,18 +131,59 @@ def self.reset_logging
       Logging.reset
     end
 
-    def self.warn_once(type, msg)
+    # Issues a warning unless the user has suppressed it.
+    #
+    def self.warn(id:, msg:, src: self)
       @mutex.synchronize do
-        @logger ||= Bolt::Logger.logger(self)
-        if @warnings.add?(type)
-          @logger.warn(msg)
-        end
+        return if @disable_warnings.include?(id)
+        Bolt::Logger.logger(src).warn("#{msg} [ID: #{id}]")
+      end
+    end
+
+    # Issues a warning unless the user has suppressed it, or a warning with the
+    # same ID has already been issued.
+    #
+    def self.warn_once(id:, msg:, src: self)
+      @mutex.synchronize do
+        return if @disable_warnings.include?(id) || !@warnings.add?(id)
+        Bolt::Logger.logger(src).warn("#{msg} [ID: #{id}]")
+      end
+    end
+
+    # Issues a deprecation warning unless the user has suppressed it. Submits
+    # analytics for the deprecation regardless of whether the warning is
+    # suppressed.
+    #
+    def self.deprecation_warning(id:, msg:, src: self)
+      @analytics&.event('Warn', 'deprecation', label: id)
+      warn(id: id, msg: msg, src: src)
+    end
+
+    # Queues a log message to be issued at a later time. This is used when
+    # parsing the command and loading the project/config, as the logger is not
+    # configured until after that happens.
+    #
+    def self.queue_message(type:, msg:, id: nil, src: nil)
```

Comment:
I don't think 'id' can ever be nil.

Reply:
https://github.com/puppetlabs/bolt/pull/2568/files#diff-69231076f18e469d86a14ebeb845fecfaec1d2adf4543f5513363c8123b051c3R51-R56